bigflow/build/spec.py

```diff
@@ -0,0 +1,262 @@
+
+
+
+@functools.lru_cache(maxsize=None)
+def get_project_spec(dir: Path = None):
+    """Reads project spec from `setup.py` and/or `pyproject.toml`.
+    Memoize results (key = project path).
+    """
+
+    dir = dir or Path.cwd()
+    return read_project_spec(dir)
+
+
+def _mabye_read_pyproject(dir: Path):
+    pyproject_toml = dir / "pyproject.toml"
+    if pyproject_toml.exists():
+        logger.info("Load config %s", pyproject_toml)
+        data = toml.load(pyproject_toml)
+        return data.get('bigflow-project')
+    else:
+        logger.debug("File %s not found", pyproject_toml)
+
+
+def read_project_spec_nosetuppy(project_dir, **kwargs):
+    """Read project spec from pyproject.toml, allowing to overwrite any options.
+    Does *NOT* inspect `setup.py` (as it is intented to be used from setup.py)"""
+
+    data = {}
+    data.update(_mabye_read_pyproject(project_dir) or {})
+    data.update(kwargs)
+    return parse_project_spec(project_dir, **data)
+
+
+def read_project_spec(dir: Path) -> BigflowProjectSpec:
+    """Reads project spec from `setup.py` and/or `pyproject.toml`"""
+
+    setuppy = dir / "setup.py"
+    prj_setuppy = dir / "project_setup.py"  # legacy
+    if not setuppy.exists() and prj_setuppy.exists():
+        logger.info("Use deprecated 'project_setup.py' instead of 'setup.py'")
+        setuppy = prj_setuppy
+
+    if setuppy.exists():
+        logger.debug("Read project spec from `setup.py` and `pyproject.toml`")
+        setuppy_kwargs = bigflow.build.dev.read_setuppy_args(setuppy)
+    else:
+        logger.debug("Read project spec only from `pyproject.toml`")
+        setuppy_kwargs = {}
+
+    try:
+        return read_project_spec_nosetuppy(dir, **setuppy_kwargs)
+    except Exception:
+        raise ValueError(
+            "The project configuration is invalid. "
+            "Check the documentation how to create a valid `setup.py`: "
+            "https://github.com/allegro/bigflow/blob/master/docs/build.md"
+        )
+
+
+# Provide defaults for project-spec
+
+def discover_project_packages(project_dir: Path):
+    ret = setuptools.find_packages(where=project_dir, exclude=["test.*", "test"])
+    logger.info(
+        "Automatically discovered %d packages: \n%s",
+        len(ret),
+        "\n".join(map(" - %s".__mod__, ret)),
+    )
+    return ret
+
+
+def read_project_requirements(project_requirements_file):
+    logger.info("Read project requirements from %s", project_requirements_file)
+    req_txt = Path(project_requirements_file)
+    recompiled = bigflow.build.pip.maybe_recompile_requirements_file(req_txt)
+    if recompiled:
+        logger.warning(textwrap.dedent(f"""
+            !!! Requirements file was recompiled, you need to reinstall packages.
+            !!! Run this command from your virtualenv:
+            pip install -r {req_txt}
+        """))
+    bigflow.build.dataflow.dependency_checker.check_beam_worker_dependencies_conflict(req_txt)  # XXX
```

Comment:
redundant comment

Reply:
not only redundant, but also a duplicate